content/posts/kubernetes-chargeback.md

```diff
@@ -0,0 +1,190 @@
+
+## Most Common Optimization Opportunities
+
+### Non-Production Environments
+
+Development and testing environments are often responsible for a significant amount of your cost, especially when they're used the same way as with traditional virtual machines or bare-metal hardware, where you often had dedicated instances running for each of those environments at all times. Some reasons in many cases were, that provisioning such an environment would traditionally take days or weeks and sharing compute resources was difficult and impractical.
+
+Kubernetes forces your system and software engineers to package applications in a way that they can be provisioned very quickly and managing compute resources is one of Kubernetes' core features. Accordingly, you can encourage them, to spin up non-production environments only when needed and implement auto-shutdowns where possible. In cases where a non-production environment needs to be running continuously, it often makes sense to reduce its memory and CPU allocation or run with a fewer instances than usual. Scaling up for load-testing and deployment tests only when needed.
+
+### Rightsizing Your Workloads
+
+Kubernetes requests and limits are a way of giving a workload such as a pod or container a guaranteed amount of resources and an upper burst limit. These values are often arbitrarily set and not by empirical evidence based on experience or load tests. For example, you may encounter pods with a whole CPU assigned in their requests when they rarely even burst to 10% of that. Rightsizing these workloads may free up a lot of resources and therefore reduce cost significantly.
+
+The proposed tools allow you to compare the resources effectively used by workloads to what amount is dedicated to them and pass this information on to the development teams.
+
+Prometheus allows you to find resource usage by percentiles, which are a good indication for sizing your pods and containers. For example, if you have a pod where the 99th percentile for CPU usage is 18% (it uses less than 18% of a CPU core 99% of the time), you may want to set requests for that pod at around 180 millicores. The best ratio or percentile to use for that indication depends on the specific workload. Load tests are recommended to find out the best values for requests and limits and for validating that applications work well when these settings have changed.
+
+> #### Careful With Memory Limits
+>
+> As opposed to CPU, memory is an incompressible resource. This means that Kubernetes cannot take away memory from a process other than killing it completely, which is not the case when the CPU is congested.
+>
+> This means, that you should set requests and limits differently for memory than for CPU. For example, if your process uses 500 MB memory at the 99th percentile, you must set memory requests to higher than that, e.g. to 750 MB, to avoid having it killed regularly.
```

Comment:
```suggestion
> This means, that you should set requests and limits differently for memory than for CPU. For example, if your process uses 500 MB memory at the 99th percentile, you must set memory requests to a value higher than 500MB, to avoid having it killed regularly.
```